Feature/xinnanli/account name format

weizsforg opened the pull request:
# Critical Changes
# Changes
- We now support customizing the format of Administrative Account and Household Account names. In HEDA Settings, on the System tab, choose a preset naming option or create your own custom naming convention. For details, see [Customize Administrative and Household Account Names](https://powerofus.force.com/articles/Resource/HEDA-Customize-Administrative-and-Household-Account-Names). Thank you, Jennifer Sanchez (Attain), for promoting this idea in the Power of Us Hub! https://powerofus.force.com/hub-ideas/1158

   If you don't take the actions described below, Administrative Account names continue to be automatically updated whenever the Contact's last name changes. And Household Account names are not automatically updated unless you enable the new setting, **Automatic Household Naming**, which renames the Household Account whenever a new Contact is added or an existing Contact is renamed, deleted, or moved to another Account.

    **Action required:**
    For the Administrative Account Name Format field, select the default value that you'd like to use. Changing the default name format does not retroactively update existing Administrative Account names.

    For the Household Account Name Format field, select the default value that you'd like to use. Decide whether to enable Automatic Household Naming. If you leave the setting disabled, when a Household Account is created, we continue to _name_ it using the default name format you've specified but we do not _rename_ the Account when Contacts within the Household change. Whether automatic naming is enabled or disabled, changing the default name format does not retroactively update existing Household Account names. Note: We don't currently support the use of custom fields in custom name formats for Household Accounts.

# Issues Closed
Fixes #218
# New Metadata
CMP_Acc_Naming_Dropdown
Hierarchy_Settings__c.Admin_Account_Naming_Format__c
Hierarchy_Settings__c.Admin_Other_Name_Setting__c
Hierarchy_Settings__c.Household_Account_Naming_Format__c
Hierarchy_Settings__c.Household_Other_Name_Setting__c
None
acctNamingOther
adminAccNameFormat
firstNameLastNameAdminACC
firstNameLastNameFamily
firstNameLastNameHH
hhAccNameFormat
hhFamily
lastNameAdminAcc
lastNameFamily
lastNameFirstNameAdminAcc
lastNameFirstNameFamily
lastNameFirstNameHH
lastNameHH
# Deleted Metadata
# Testing Notes
1. Deploy the HEAD package to Dev org or Scratch org
2. Go to HEDA setting page and switch to system tab
3. Select Admin account or Household account as default record type
4. Select naming setting for admin account and household account
4. Create a contact and make sure the name of account is correct

weizsforg commented on src/classes/ACCT_IndividualAccounts_TDTM.cls:
```diff
@@ -142,6 +153,14 @@ public class ACCT_IndividualAccounts_TDTM extends TDTM_Runnable {
+                            if (c.Account.RecordTypeId == UTIL_Describe.getHhAccRecTypeID()
+                                    && UTIL_CustomSettingsFacade.getSettings().Automatic_Household_Naming__c != false){
+                                setAccountIdHHToRename.add(c.AccountId);
+                            }
```
I think in here, we also need to update oldContacts[i].Account name if oldContacts[i].AccountId != null.

xinnanlisf replied:
Good catch!

weizsforg commented on src/aura/STG_CMP_System/STG_CMP_System.cmp:
```diff
@@ -270,5 +275,94 @@
+            </aura:if>
+        </div>
+        
+        <div class="slds-col slds-size--1-of-2" >
```
I think there is a small UI issue here. System does not show the current setting. 

Steps to reproduce:
1. Deploy Master branch to the org
2. Deploy the Account Naming branch to the org
3. Go to Naming Setting, click Edit and click Save
4. The Naming settings in the UI shows blank. But the value actually saved in the custom settings

xinnanlisf replied:
Sorry which setting are you referring to? I can't seem to reproduce it.

weizsforg commented:
@xinnanlisf Please have a look about the comments in this PR. Also, looks like this branch is out of date. Can you update this branch?

weizsforg commented:
@xinnanlisf Maybe we should ask someone to review the 'Changes' in this PR?

xinnanlisf commented:
@lianeyee 

We might need to update the release notes. Household accounts will now be updated if a child contact is deleted or moved to another account(This account will also be updated!).

lianeyee commented:
We definitely need to clean up the release notes for public consumption. It's on my radar for today. Thanks, @xinnanlisf

lianeyee commented:
@xinnanlisf @weizsforg When you have a chance, please review the revised release notes. I'm afraid they got a bit long and involved. But I think it's better to be as clear and specific as possible about how Household Account naming behaves.

lianeyee commented:
@xinnanlisf It just occurred to me that the help text for Automatic Household Naming might be  incorrect now. 

According to the WI, it said: "Enables automatic naming of Household Accounts, based on the names of Contacts in the Household. If enabled, we also automatically rename Household Accounts when new Contacts are added or existing Contacts are renamed."

According to the behavior now, shouldn't it say something like this? "Enables automatic renaming of Household Accounts, based on the names of Contacts in the Household. If enabled, we automatically rename Household Accounts when a new Contact is added to the Household or an existing Contact is renamed, deleted, or moved to another Household."

xinnanlisf commented:
@lianeyee 

Good catch. I will go ahead and make the changes. :)

weizsforg commented:
@xinnanlisf @lianeyee I think there are still some questions when the Contact is renamed, deleted, or moved to another Account for both Admin and HH Account model.  

Currently:
1. For Admin Account Model: We only update Account Name when LastName of the Contact changes. We do not update Account Name when Contact is renamed, deleted, or moved to another Account.

2. For HH Account Model: We included the logic to handle Contact is renamed, deleted, or moved to another Account in this PR. But I think there is an issue when we move the Contact from One Account to Another Account.

It will be good to document the expect behavior (maybe with an example) when Contact is renamed, deleted, or moved to another Account for both Admin and HH Account model in HEDA. 

Any ideas?

xinnanlisf commented:
@weizsforg 

Are you questioning about the help text? I think it is used only for Household account, correct?

weizsforg commented:
@xinnanlisf Actually, it is about the actual logic. 

In the "Changes" of this PR and the Quip doc, we mention "Administrative Account names continue to be automatically updated whenever a new Contact is added to an Account or an existing Contact is renamed, deleted, or moved to another Account. ". 

But I don't think we currently doing that. We only update Admin Acc Name when Contact LastName changes.

lianeyee commented:
That's my error then. For Admin Account, it sounds like it should be "Administrative Account names continue to be automatically updated whenever the Contact's name changes." (If they choose a name format that includes First Name, we shouldn't specify just Last Name, right?)

weizsforg commented:
I think the code logic only handle the Contact LastName changes of the Admin Account model.  What do you think? @xinnanlisf Should we just mention Name instead of LastName?

weizsforg commented:
Thanks for all the great work! @xinnanlisf
Looks like I can not be the reviewer since this PR is created by me. Can you approve it @xinnanlisf ? 

@annejackson This PR is ready for QA now. I think one important thing to test is to make sure the default behavior does not change when user do nothing after we ship this feature.

annejackson commented:
Hey @xinnanlisf - I'm not ready to test (or pass) this yet, but I just noticed that the "merge pull request" button is disabled...do we know why?

xinnanlisf commented:
@annejackson 

That's mostly because @weizsforg reverted the pull request and someone else need to approve it. I just approved it and it should be ready for merge.

annejackson commented:
Yep, I see that it's available to merge now. Thanks @xinnanlisf

annejackson commented:
@xinnanlisf saw this issue while using the Other feature. See screenshots. 
![settings when saw bug](https://user-images.githubusercontent.com/35311289/36555681-b0b02930-17d0-11e8-97a5-f1b780269561.png)
![bug when adding contact with other settings](https://user-images.githubusercontent.com/35311289/36555682-b0c11ce0-17d0-11e8-8e45-1e6a7be62436.png)

xinnanlisf commented:
@annejackson 

Should be good now. Good catch! I would suggest let @weizsforg do another code review before merging it. :)

annejackson commented:
OK @xinnanlisf I can move into my other test cases that I have left using my current scratch org.  @ceiroa  since Wei is out on PTO, can you or someone else do a review, or how does this normally work?

ceiroa commented:
@annejackson @xinnanlisf Yes, I'll take a look.

annejackson commented:
@xinnanlisf  here's the info, screenshots I told you about earlier. 
![settings for bug2](https://user-images.githubusercontent.com/35311289/36563307-b46d3bd2-17e7-11e8-8c4b-5b3b3f837fb6.png)
![bug2](https://user-images.githubusercontent.com/35311289/36563308-b47f1168-17e7-11e8-9288-0e9c3da14627.png)
fyi @ceiroa

xinnanlisf commented:
@annejackson  should be good now! Thanks for your help.

annejackson commented:
@ceiroa  can you review the changes? Thanks!

ceiroa commented:
@annejackson The changes have been reviewed, and this is all ready for one last review next Monday. Thanks!

xinnanlisf commented:
@lianeyee 
I just updated the release notes. Could you do a quick review please? Also, regarding the document you have been working on, we might need to add a note in it as well. Let me know if you need any help or any questions.

lianeyee commented:
Thanks for updating the release note, @xinnanlisf. I moved that item down to the Household Account section because this limitation applies only to Household Accounts. I will add a similar note to the documentation.

annejackson commented:
Testing has passed in the scratch org. Merging this pull request. Nice work @xinnanlisf. 
Fyi @ceiroa
